# Group Project 1

## Problem1 Words Count

### Part A
#### How many times are each of the following characters mentioned by name in the text of The Hound of the Baskervilles?

In [1]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt

--2020-09-25 23:25:15--  https://s3.amazonaws.com/dmfa-2020/project-1/hound.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.12.54
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.12.54|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 345659 (338K) [text/plain]
Saving to: ‘hound.txt’

hound.txt           100%[===================>] 337.56K  --.-KB/s    in 0.008s  

2020-09-25 23:25:16 (41.1 MB/s) - ‘hound.txt’ saved [345659/345659]



In [1]:
! grep -w -o -i Holmes hound.txt | wc -l

193


In [2]:
! grep -w -o -i Watson hound.txt | wc -l

114


In [3]:
! grep -w -o -i Barrymore hound.txt | wc -l

73


In [4]:
! grep -w -o -i  Mortimer hound.txt | wc -l

90


* Answer: There are 193 Holmes, 114 Walton, 73 Barrymore and 90 Mortimer in the text of the Hound of the Baskervilles.

* We use !grep to search for words, -w for limiting the names we want to match to word boundaries, -o for printing only matched parts of a maching line, withe each such part on a separate output line, -i for lowering case all words and -l for displaying the number of words "Holmes", "Watson", "Barrymore", "Mortimer".

### Part B
#### How many times do each of the characters Hamlet, Polonius, Ophelia, and Horatio have speaking lines in Hamlet? Keep in mind that this is the text of a play.

In [2]:
!wget https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt

--2020-09-25 23:25:20--  https://s3.amazonaws.com/dmfa-2020/project-1/hamlet.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.185.61
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.185.61|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 204885 (200K) [text/plain]
Saving to: ‘hamlet.txt’

hamlet.txt          100%[===================>] 200.08K  --.-KB/s    in 0.006s  

2020-09-25 23:25:20 (34.6 MB/s) - ‘hamlet.txt’ saved [204885/204885]



In [8]:
!grep -o -w -F "HAMLET." hamlet.txt | wc -l

360


In [9]:
! grep -o -w -F "POLONIUS." hamlet.txt | wc -l

86


In [10]:
! grep -o -w -F "OPHELIA." hamlet.txt | wc -l

58


In [11]:
! grep -o -w -F "HORATIO." hamlet.txt | wc -l

107


* Answer: There are 360 times of character Hamlet, 86 times of character Polonius, 58 times of character Ophelia and 107 times of Horatio having speaking lines in Hamlet. 

In [12]:
!grep -o -w -F "HAMLET." hamlet.txt

HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.
HAMLET.


* The last command shows that we can use the command to get the write uppercased name with a dot so that we can know which sentences are the characters' speaking lines.

* We use -F to fix the string we want to search and interpret 'the name with a dot' as a list of fixed string, separate by newline, any of which is to be matched

* Reference:  https://www.explainshell.com/explain?cmd=grep+-o+-w+-F+%22HAMLET.%22+hamlet.txt%7C+wc+-l

        

## Problem 2 - Capital Bikeshare

### Part A

In [1]:
! wget https://s3.amazonaws.com/dmfa-2020/project-1/2018-capitalbikeshare-tripdata.zip

--2020-09-27 14:54:50--  https://s3.amazonaws.com/dmfa-2020/project-1/2018-capitalbikeshare-tripdata.zip
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.177.77
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.177.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18365732 (18M) [application/zip]
Saving to: ‘2018-capitalbikeshare-tripdata.zip’

2018-capitalbikesha 100%[===================>]  17.51M  93.9MB/s    in 0.2s    

2020-09-27 14:54:51 (93.9 MB/s) - ‘2018-capitalbikeshare-tripdata.zip’ saved [18365732/18365732]



In [2]:
! unzip 2018-capitalbikeshare-tripdata.zip

Archive:  2018-capitalbikeshare-tripdata.zip
  inflating: 201807-capitalbikeshare-tripdata.csv  
  inflating: 201808-capitalbikeshare-tripdata.csv  


#### 1. Unzip the data and combine the two CSV files using csvstack.

In [3]:
! csvstack 201807-capitalbikeshare-tripdata.csv 201808-capitalbikeshare-tripdata.csv > biketrip.csv

In [4]:
! head biketrip.csv | csvlook

| Duration |          Start date |            End date | Start station number | Start station                  | End station number | End station                                  | Bike number | Member type |
| -------- | ------------------- | ------------------- | -------------------- | ------------------------------ | ------------------ | -------------------------------------------- | ----------- | ----------- |
|      267 | 2018-07-01 00:00:58 | 2018-07-01 00:05:26 |               31,608 | 8th & Eye St SE / Barracks Row |             31,610 | Eastern Market / 7th & North Carolina Ave SE | W23296      | Member      |
|      530 | 2018-07-01 00:01:00 | 2018-07-01 00:09:50 |               31,230 | Metro Center / 12th & G St NW  |             31,266 | 11th & M St NW                               | W20838      | Member      |
|    3,223 | 2018-07-01 00:01:07 | 2018-07-01 00:54:50 |               31,620 | 5th & F St NW                  |             31,289 | Henry Bacon Dr & Lincoln Memor

In [21]:
!csvcut -n biketrip.csv

  1: Duration
  2: Start date
  3: End date
  4: Start station number
  5: Start station
  6: End station number
  7: End station
  8: Bike number
  9: Member type


#### 2. Which 10 Capital Bikeshare stations were the most popular departing stations in July and August 2018?

In [5]:
! csvcut -c 4,5 biketrip.csv > populardeparting.csv

In [6]:
! sort -rn populardeparting.csv | uniq -c| sort -rn| head -10

  15474 31258,Lincoln Memorial
  13082 31623,Columbus Circle / Union Station
  11999 31247,Jefferson Dr & 14th St SW
  11671 31289,Henry Bacon Dr & Lincoln Memorial Circle NW
  11650 31288,4th St & Madison Dr NW
  10357 31248,Smithsonian-National Mall / Jefferson Dr & 12th St SW
   9554 31249,Jefferson Memorial
   8620 31290,17th St & Independence Ave SW
   8583 31200,Massachusetts Ave & Dupont Circle NW
   8569 31201,15th & P St NW


* Answer: We use !csvcut to combine the start station number and name column into a new file called populardeparting.csv. Then we use !sort to sort the file and choose the station number which is repeated for many time. Each station number just stands for one station name. The most 10 popular departing stations are shown above.

#### Which 10 stations were the most popular destination stations in July and August 2018?

In [7]:
! csvcut -c 6,7 biketrip.csv > populardestination.csv

In [8]:
! sort populardestination.csv |uniq -c| sort -rn| head -10

  15642 31258,Lincoln Memorial
  13635 31623,Columbus Circle / Union Station
  12135 31247,Jefferson Dr & 14th St SW
  11722 31289,Henry Bacon Dr & Lincoln Memorial Circle NW
  11555 31288,4th St & Madison Dr NW
  10693 31248,Smithsonian-National Mall / Jefferson Dr & 12th St SW
   9866 31249,Jefferson Memorial
   9141 31200,Massachusetts Ave & Dupont Circle NW
   8884 31201,15th & P St NW
   8640 31290,17th St & Independence Ave SW


* Answer: We use !csvcut to combine the end station number and name column into a new file called populardestination.csv. Then we use !sort to sort the file and choose the station number which is repeated for many time. Each station number just stands for one station name. The most 10 popular destination stations are shown above.

## Part B
#### 1. For the most popular departure station, which 10 bikes were used most in trips departing from there?

In [36]:
!csvgrep -c5 -r "^Lincoln Memorial" biketrip.csv | csvcut -c4-5,8 | csvgrep -c3 -r "^W" | sort | uniq -c | sort -rn | head 

     18 31258,Lincoln Memorial,W01311
     17 31258,Lincoln Memorial,W22553
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W23074
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W21882
     15 31258,Lincoln Memorial,W20527
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23003
     14 31258,Lincoln Memorial,W22369
sort: write failed: 'standard output': Broken pipe
sort: write error


In [10]:
!csvgrep -c 4 -r "31258" biketrip.csv | csvcut -c 4,5,8 | csvgrep -c3 -r "^W" | sort | uniq -c | sort -rn | head 

     18 31258,Lincoln Memorial,W01311
     17 31258,Lincoln Memorial,W22553
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W23074
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W21882
     15 31258,Lincoln Memorial,W20527
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23003
     14 31258,Lincoln Memorial,W22369
sort: write failed: 'standard output': Broken pipe
sort: write error


* Answer: We use two command lines to find the most 10 bikes used from the departing station. The 1st command provides the station Lincoln Memorial which has the top 10 bikes used but there is other station which also includes the term "Lincoln Memorial", so we use the 2nd command line to find the repeated station number to make sure that the results are correct.

* For the most popular departure station, the most 10 bikes are shown above


#### 2. Which 10 bikes were used most in trips ending at the most popular destination station?

In [37]:
!csvgrep -c7 -r "^Lincoln Memorial" biketrip.csv | csvcut -c6-7,8 | csvgrep -c3 -r "^W" | sort | uniq -c | sort -rn | head 

     17 31258,Lincoln Memorial,W22553
     17 31258,Lincoln Memorial,W01311
     16 31258,Lincoln Memorial,W21882
     16 31258,Lincoln Memorial,W20527
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W22369
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23074
     14 31258,Lincoln Memorial,W23003
sort: write failed: 'standard output': Broken pipe
sort: write error


In [11]:
!csvgrep -c 6 -r "31258" biketrip.csv | csvcut -c 6,7,8 | csvgrep -c3 -r "^W" | sort | uniq -c | sort -rn | head 

     17 31258,Lincoln Memorial,W22553
     17 31258,Lincoln Memorial,W01311
     16 31258,Lincoln Memorial,W21882
     16 31258,Lincoln Memorial,W20527
     16 31258,Lincoln Memorial,W00919
     15 31258,Lincoln Memorial,W22567
     15 31258,Lincoln Memorial,W22369
     14 31258,Lincoln Memorial,W23311
     14 31258,Lincoln Memorial,W23074
     14 31258,Lincoln Memorial,W23003
sort: write failed: 'standard output': Broken pipe
sort: write error


* Answer: We use two command lines to find the most 10 bikes used from the destination station. The 1st command provides the station Lincoln Memorial which has the top 10 bikes used but there is other station which also includes the term "Lincoln Memorial", so we use the 2nd command line to find the repeated station number to make sure that the results are correct.

* For the most popular destination station, the most 10 bikes are shown above

## Problem 3 Filters

### Part A

#### 1. Demonstrate a pipeline that performs a count of the top ten most common words in The Hound of the Baskervilles.

In [13]:
!grep -oE '\w{{1,}}' hound.txt \
    | tr '[:upper:]' '[:lower:]' \
    | sort \
    | uniq -c \
    | sort -rn \
    | head -10

   3515 the
   1716 of
   1696 and
   1501 i
   1488 to
   1369 a
   1159 that
   1025 it
    968 in
    922 he
sort: write failed: 'standard output': Broken pipe
sort: write error


In [12]:
!cat hound.txt | tr -sc '[:alpha:]' '[\n*]' | tr '[:upper:]' '[:lower:]' | sort | uniq -c | sort -rn | head -10

   3515 the
   1716 of
   1696 and
   1501 i
   1488 to
   1369 a
   1159 that
   1025 it
    968 in
    922 he
sort: write failed: 'standard output': Broken pipe
sort: write error


* Answer: There are two ways to create a pipeline that performs a count of the top ten most common words in The Hound of The Baskervilles. 

#### 2. Write your own lowercase filter that replaces tr [:upper:] [:lower:] to transform text into lower case.

In [3]:
!chmod +x lowercase.py

In [6]:
!cat hound.txt | ./lowercase.py | head

project gutenberg’s the hound of the baskervilles, by a. conan doyle

this ebook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  you may copy it, give it away or
re-use it under the terms of the project gutenberg license included
with this ebook or online at www.gutenberg.org


title: the hound of the baskervilles

Traceback (most recent call last):
  File "./lowercase.py", line 10, in <module>
    print(txt.lower())
BrokenPipeError: [Errno 32] Broken pipe
cat: write error: Broken pipe


In [41]:
!cat hound.txt | ./lowercase.py > lowercasehound.txt

cat: write error: Broken pipe


* Answer: We created a python file named lowercase.py which contains codes to make the text in the file to become lower case

#### 3. With your new filter, repeat the original pipeline, and substitute your new filters as appropriate.

In [7]:
!cat hound.txt | ./lowercase.py | grep -oE '\w{{1,}}'| sort |uniq -c | sort -rn|head

cat: write error: Broken pipe
   3515 the
   1716 of
   1696 and
   1501 i
   1488 to
   1369 a
   1159 that
   1025 it
    968 in
    922 he
sort: write failed: 'standard output': Broken pipe
sort: write error


### Part B
### 1. Write a filter that removes at least ten common words of English text, commonly known as “stop words”. 
#### We searched stopwords online and choose almost 20 stopwords and already generated stopwords list in the simplefilter.py. We will filter the text to exclude these words. 

In [47]:
!head lowercasehound.txt

project gutenberg’s the hound of the baskervilles, by a. conan doyle

this ebook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  you may copy it, give it away or
re-use it under the terms of the project gutenberg license included
with this ebook or online at www.gutenberg.org


title: the hound of the baskervilles



In [48]:
! touch filteredhound.txt

In [49]:
!chmod +x simplefilter.py

In [50]:
!cat lowercasehound.txt | ./simplefilter.py > filteredhound.txt

cat: write error: Broken pipe


In [51]:
!head filteredhound.txt

['project', 'gutenberg’s', 'hound', 'baskervilles,', 'a.', 'conan', 'doyle']
[]
['this', 'ebook', 'use', 'anyone', 'anywhere', 'cost', 'with']
['almost', 'restrictions', 'whatsoever.', 'may', 'copy', 'it,', 'give', 'it', 'away', 'or']
['re-use', 'it', 'under', 'terms', 'project', 'gutenberg', 'license', 'included']
['with', 'this', 'ebook', 'or', 'online', 'www.gutenberg.org']
[]
[]
['title:', 'hound', 'baskervilles']
[]


* Answer: We used the filter to delete the stopwords (the, of, by, is, for, of, at, no, and, you) from hound.txt and we saved it into a new file, filteredhound.txt. This step can be prepared for the following question.

### 2. Add your stop word filter to a word count pipeline and show the most common 25 words in The Hound of the Baskervilles when stop words are removed. 

In [52]:
!chmod +x simplefilter2.py

In [53]:
!cat filteredhound.txt | ./simplefilter2.py

'to' 1475
'a' 1342
'i' 1283
'that' 1098
'in' 953
'he' 879
'it' 867
'was' 799
'his' 683
'have' 539
'had' 503
'with' 484
'my' 451
'we' 436
'which' 424
'as' 412
'not' 376
'this' 364
'but' 351
'be' 348
'sir' 320
'upon' 315
'me' 298
'from' 296
'him' 294
cat: write error: Broken pipe


Answer: The most common 25 words in The Hound of the Baskervilles when stop words are removed are shown above.

#### Reference: https://stackoverflow.com/questions/21722955/counting-words-in-a-file-using-counter-object 